File: algorithms/graphs/J.cpp
#include <cstdio>
#include <vector>
#include <tuple>
#include <algorithm>

/**
 * @author Saveliy Bakturin
 * <p>
 * Don't write off, if you don't wanna be banned!
 */

class dsu {
private:
	std::vector<int> root;
	std::vector<int> rank;

	void make(int const& v) {
		root[v] = v;
		rank[v] = 0;
	}

public:
	dsu(std::size_t const &size) :
			root(std::vector<int>(size + 1)),
			rank(std::vector<int>(size + 1)) {
		for (std::size_t i = 1; i < size + 1; i++) make(static_cast<int>(i));
	}

	int find(int const& v) {
		if (v == root[v]) return v;
		else return root[v] = find(root[v]);
	}

	void unification(int const& l, int const& r) {
		int a = find(l);
		int b = find(r);
		if (a != b) {
			if (rank[a] < rank[b]) std::swap(a, b);
			root[b] = a;
			if (rank[a] == rank[b]) rank[a]++;
		}
	}
};

int main() {
	using node_t = std::tuple<int, int, int>;
	std::size_t n, m;
	std::scanf("%zu %zu", &n, &m);
	dsu dsu(n);
	std::vector<node_t> edges;
	long long sum = 0;
	for (std::size_t i = 0; i < m; i++) {
		int u, v, w;
		std::scanf("%i %i %i", &u, &v, &w);
		edges.emplace_back(u, v, w);
	}
	std::sort(edges.begin(), edges.end(), [](node_t a, node_t b){
		return std::get<2>(a) < std::get<2>(b);
	});
	for (auto const& item : edges) {
		if (dsu.find(std::get<0>(item)) != dsu.find(std::get<1>(item))) {
			sum += std::get<2>(item);
			dsu.unification(std::get<0>(item), std::get<1>(item));
		}
	}
	std::printf("%lld", sum);
}

File: algorithms/half/J/J.java
import java.util.Arrays;
import java.util.Scanner;

/**
 * @author Saveliy Bakturin
 * <p>
 * Don't write off, if you don't wanna be banned!
 */

public class J {
	public static void main(final String... args) {
		final Scanner in = new Scanner(System.in);
		final int n = in.nextInt();
		in.nextLine();
		final String[] s = new String[n];
		for (int i = 0; i < n; i++) {
			s[i] = in.nextLine();
		}
		in.close();
		Arrays.sort(s);
		String[] prev_s = new String[1024];
		int prev_l = 0;
		for (int i = 0; i < n; i++) {
			final String[] sp = s[i].split("/");
			if (sp.length == 1) {
				System.out.println(s[i]);
				continue;
			}
			if ((sp.length > prev_l) || (sp.length > 1 && sp.length < prev_l)) {
				System.out.println(" ".repeat(2 * (sp.length - 1)) + sp[sp.length - 1]);
			}
			if (sp.length == prev_l) {
				int count = 0;
				for (int j = 0; j < prev_l; j++) {
					if (prev_s[j].compareTo(sp[j]) == 0) {
						count++;
					}
				}
				if (count == prev_l - 1) {
					System.out.println(" ".repeat(2 * (sp.length - 1)) + sp[sp.length - 1]);
				}
			}
			prev_s = sp;
			prev_l = sp.length;
		}
	}
}
